src/main/java/org/junitpioneer/jupiter/EnvironmentVariableExtension.java

```diff
@@ -0,0 +1,167 @@
+		Set<String> variablesToClear;
+		Map<String, String> variablesToSet;
+		try {
+			variablesToClear = findRepeatableAnnotations(context, ClearEnvironmentVariable.class)
+					.stream()
+					.map(ClearEnvironmentVariable::key)
+					.collect(Utils.distinctToSet());
+			variablesToSet = findRepeatableAnnotations(context, SetEnvironmentVariable.class)
+					.stream()
+					.collect(toMap(SetEnvironmentVariable::key, SetEnvironmentVariable::value));
+			preventClearAndSetSameEnvironmentVariables(variablesToClear, variablesToSet.keySet());
+		}
+		catch (IllegalStateException ex) {
+			throw new ExtensionConfigurationException("Don't clear/set the same environment variable more than once.",
+				ex);
+		}
+
+		storeOriginalEnvironmentVariables(context, variablesToClear, variablesToSet.keySet());
+		reportWarning(context);
+		EnvironmentVariableUtils.clear(variablesToClear);
+		EnvironmentVariableUtils.set(variablesToSet);
+	}
+
+	private <A extends Annotation> List<A> findRepeatableAnnotations(ExtensionContext context,
+			Class<A> annotationType) {
+		return context
+				.getElement()
+				.map(element -> AnnotationSupport.findRepeatableAnnotations(element, annotationType))
+				.orElseGet(Collections::emptyList);
+	}
+
+	private void preventClearAndSetSameEnvironmentVariables(Collection<String> variablesToClear,
+			Collection<String> variablesToSet) {
+		variablesToClear
+				.stream()
+				.filter(variablesToSet::contains)
+				.reduce((k0, k1) -> k0 + ", " + k1)
+				.ifPresent(duplicateKeys -> {
+					throw new IllegalStateException(
+						"Cannot clear and set the following environment variable at the same time: " + duplicateKeys);
+				});
+	}
+
+	private void storeOriginalEnvironmentVariables(ExtensionContext context, Collection<String> clearVariables,
+			Collection<String> setVariables) {
+		context.getStore(NAMESPACE).put(BACKUP, new EnvironmentVariableBackup(clearVariables, setVariables));
+	}
+
+	private void reportWarning(ExtensionContext context) {
+		boolean wasReported = REPORTED_WARNING.getAndSet(true);
+		if (!wasReported)
+			context.publishReportEntry(WARNING_KEY, WARNING_VALUE);
+	}
+
+	@Override
+	public void afterEach(ExtensionContext context) {
+		boolean present = Utils
+				.annotationPresentOnTestMethod(context, ClearEnvironmentVariable.class, ClearEnvironmentVariables.class,
+					SetEnvironmentVariable.class, SetEnvironmentVariables.class);
+		if (present) {
+			restoreOriginalEnvironmentVariables(context);
+		}
+	}
+
+	@Override
+	public void afterAll(ExtensionContext context) {
+		restoreOriginalEnvironmentVariables(context);
+	}
+
+	private void restoreOriginalEnvironmentVariables(ExtensionContext context) {
+		context.getStore(NAMESPACE).get(BACKUP, EnvironmentVariableBackup.class).restoreVariables();
+	}
+
+	// this method is needed by the tests that verify whether the warning is correctly reported
+	static void resetWarning() {
+		REPORTED_WARNING.set(false);
+	}
```

Comment:
A bit weird to add this method just for tests. What do you think?

Reply:
I don't like methods in prod code just for tests.

Reply:
Me neither. Fixed it by making the field accessible, so this can be done directly by the tests. (Doesn't really change anything functionally, but not having that method there just feels better. :grin: )